[SPARK-20641][core] Add key-value store abstraction and LevelDB implementation.

vanzin opened the pull request:
This change adds an abstraction and LevelDB implementation for a key-value
store that will be used to store UI and SHS data.

The interface is described in KVStore.java (see javadoc). Specifics
of the LevelDB implementation are discussed in the javadocs of both
LevelDB.java and LevelDBTypeInfo.java.

Included also are a few small benchmarks just to get some idea of
latency. Because they're too slow for regular unit test runs, they're
disabled by default.

Tested with the included unit tests, and also as part of the overall feature
implementation (including running SHS with hundreds of apps).

Commit:
SHS-NG M1: Add KVStore abstraction, LevelDB implementation.

The interface is described in KVIndex.java (see javadoc). Specifics
of the LevelDB implementation are discussed in the javadocs of both
LevelDB.java and LevelDBTypeInfo.java.

Included also are a few small benchmarks just to get some idea of
latency. Because they're too slow for regular unit test runs, they're
disabled by default.

Commit:
SHS-NG M1: Add support for arrays when indexing.

This is needed because some UI types have compound keys.

Commit:
SHS-NG M1: Fix counts in LevelDB when updating entries.

Also add unit test. When updating, the code needs to keep track of
the aggregated delta to be added to each count stored in the db,
instead of reading the count from the db for each update.

Commit:
SHS-NG M1: Try to prevent db use after close.

This causes JVM crashes in the leveldb library, so try to avoid it;
if there are still issues, we'll neeed locking.

Commit:
SHS-NG M1: Use Java 8 lambdas.

Also rename LevelDBIteratorSuite to work around some super weird
issue with sbt.

Commit:
SHS-NG M1: Compress values stored in LevelDB.

LevelDB has built-in support for snappy compression, but it seems
to be buggy in the leveldb-jni library; the compression threads
don't seem to run by default, and when you enable them, there are
weird issues when stopping the DB.

So just do compression manually using the JRE libraries; it's probably
a little slower but it saves a good chunk of disk space.

Commit:
SHS-NG M1: Use type aliases as keys in Level DB.

The type name gets repeated a lot in the store, so using it as the prefix
for every key causes disk usage to grow unnecessarily. Instead, create a
short alias for the type and keep a mapping of aliases to known types in
a map in memory; the map is also saved to the database so it can be read
later.

Commit:
SHS-NG M1: Separate index introspection from storage.

The new KVTypeInfo class can help with writing different implementations
of KVStore without duplicating logic from LevelDBTypeInfo.

Commit:
SHS-NG M1: Remove unused methods from KVStore.

Turns out I ended up not using the raw storage methods in KVStore, so
this change removes them to simplify the API and save some code.

Commit:
SHS-NG M1: Add "max" and "last" to kvstore iterators.

This makes it easier for callers to control the end of iteration,
making it easier to write Scala code that automatically closes
underlying iterator resources. Before, code had to use Scala's
"takeWhile", convert the result to a list, and manually close the
iterators; with these two parameters, that can be avoided in a
bunch of cases, with iterators auto-closing when the last element
is reached.

SparkQA commented:
**[Test build #76589 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/76589/testReport)** for PR 17902 at commit [`5197c21`](https://github.com/apache/spark/commit/5197c218525db2ad849dfe77d83dddf2311bb5ad).

vanzin commented:
For context:
- Project link: https://issues.apache.org/jira/browse/SPARK-18085
- Upcoming PRs that build on this code: https://github.com/vanzin/spark/pulls

SparkQA commented:
**[Test build #76589 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/76589/testReport)** for PR 17902 at commit [`5197c21`](https://github.com/apache/spark/commit/5197c218525db2ad849dfe77d83dddf2311bb5ad).
 * This patch **fails from timeout after a configured wait of \`250m\`**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/76589/
Test FAILed.

Commit:
SHS-NG M1: Add missing test tags dependency.

SparkQA commented:
**[Test build #76605 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/76605/testReport)** for PR 17902 at commit [`63e0a58`](https://github.com/apache/spark/commit/63e0a58b01bd622d6a3f2dc8fbe72c819493c152).

SparkQA commented:
**[Test build #76605 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/76605/testReport)** for PR 17902 at commit [`63e0a58`](https://github.com/apache/spark/commit/63e0a58b01bd622d6a3f2dc8fbe72c819493c152).
 * This patch **fails from timeout after a configured wait of \`250m\`**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/76605/
Test FAILed.

vanzin commented:
Hmm... I've run into these SQL test failures in my tests, and have this workaround in my branch: https://github.com/vanzin/spark/pull/22/files#diff-edd374dbb96bc16363b65dab1e554793R114

In the meantime... retest this please.

vanzin commented:
retest this please

SparkQA commented:
**[Test build #76688 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/76688/testReport)** for PR 17902 at commit [`63e0a58`](https://github.com/apache/spark/commit/63e0a58b01bd622d6a3f2dc8fbe72c819493c152).

SparkQA commented:
**[Test build #76688 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/76688/testReport)** for PR 17902 at commit [`63e0a58`](https://github.com/apache/spark/commit/63e0a58b01bd622d6a3f2dc8fbe72c819493c152).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/76688/
Test PASSed.

Commit:
SHS-NG M1: Add child (or secondary) indices.

This change adds child indices, which allow sorting a result set
according to two indexed fields. To use a concrete example, with
this change you can efficiently retrieve a list of tasks that belong
to a particular stage attempt (the primary index), sorted by the
time they were launched (the secondary index).

I also added a more verbose description of what the LevelDB structure
looks like in LevelDBTypeInfo.java, since that actually helped me
visualize what was going on while adding the new code.

I also made a bunch of other changes targeted at speeding up this
code:
- avoid String -> byte[] conversions and other expensive operations,
  especially when creating keys based on numeric values.
- avoid calculating the same key prefix multiple times when writing.
- do not store counts in end markers when not needed; the API only
  require some of those to be stored for efficient retrieval, and
  LevelDB doesn't require a key to exist before you can "seek" to it.

Commit:
SHS-NG M1: Remove LevelDBWriteBatch.

After recent changes, LevelDBWriteBatch becomes sort of redundant; the
needed logic can be moved to LevelDBTypeInfo to save some more redundant
operations (including reads, writes, can calculation of keys), also
saving some code.

I also backported some changes from a later set, making DBIteratorSuite
an abstract class. For whatever odd reason, sbt / junit-interface have
a hard time finding all the test suites in this module, and this particular
change seems to help.

vanzin commented:
Given the overwhelming interest in reviewing this code so far, I chose to include some changes I was working on for future milestones, in order to avoid code churn later on. These include:

- some test refactoring from SPARK-20655, to work around some weird issue with sbt not running all JUnit tests in the new module
- many optimizations and a new feature I added while working on SPARK-20657.

I also included some more (javadoc) documentation to help with understanding with how data is stored in LevelDB.

SparkQA commented:
**[Test build #76948 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/76948/testReport)** for PR 17902 at commit [`85e8930`](https://github.com/apache/spark/commit/85e893021497386effff3d58b902943fe896e858).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/76948/
Test PASSed.

ajbozarth commented:
Sorry I haven't had time to review this, given it's 100% new code I haven't have time to give it a proper read through. Hopefully I can find time by end of month (if not sooner)

jsoltren commented on common/kvstore/src/test/java/org/apache/spark/kvstore/LevelDBSuite.java:
```diff
@@ -0,0 +1,291 @@
+    db = new LevelDB(dbpath);
+    assertEquals(LevelDB.STORE_VERSION,
+      db.serializer.deserializeLong(db.db().get(LevelDB.STORE_VERSION_KEY)));
+    db.db().put(LevelDB.STORE_VERSION_KEY, db.serializer.serialize(LevelDB.STORE_VERSION + 1));
```
Why the + 1 here?

vanzin replied:
Because the test is explicitly testing a mismatch in the version.

jsoltren reviewed (commented):
This is a large change. Provisionally it looks okay to me, though I'd like a bit more soak time with it before I say for sure. It seems relatively straightforward and fits in clearly with your "Next Generation Spark History Server / UI" design document. As a reminder to anyone else following along, the goal here is to provide an API for writing values to, and reading values from, a LevelDB database, so that the Spark History Server can use this instead of reading from flat log files.

I'm still having a hard time visualizing what a row in the LevelDB "looks like". 

I wonder if, perhaps, some of the documentation you provided in earlier commit messages should be propagated to comments in the code itself. My understanding is that those commit messages will be lost when this PR is merged. For example, explaining why indices can take lists of elements, or why you chose to use Java's built-in gzip compression instead of something else. There are a number of good design decisions that are not obvious from code inspection alone (unless you already know the justification behind those decisions).

I'll have more to say later as I keep reading the code but wanted to make sure to post /something/ before the long weekend.

vanzin commented on common/kvstore/src/test/java/org/apache/spark/kvstore/DBIteratorSuite.java:
```diff
@@ -0,0 +1,498 @@
+    // the same way the store is expected to.
+    CustomType1 first = allEntries.get(0);
+    clashingEntries = new ArrayList<>();
+    for (int i = 0; i < RND.nextInt(MIN_ENTRIES) + 1; i++) {
```
Because `Random.nextInt(int)` can return 0.

vanzin commented:
> I'm still having a hard time visualizing what a row in the LevelDB "looks like".

It's a key/value store, there are no "rows". The key format is explained in the javadoc for `LevelDBTypeInfo`.

> For example, explaining why indices can take lists of elements

? 

> why you chose to use Java's built-in gzip

There's no special reason why I chose that specific implementation other than it's readily available in the JRE.

In general, if there's no explanation for some specific choice I made in the code, it's generally because it doesn't really matter. For example, it doesn't matter what compression algo is used. This code is not in any critical path and applications don't even know whether the data is compressed.

If there's anything in particular you want to know, feel free to ask.

jsoltren commented:
> >For example, explaining why indices can take lists of elements

> ?

That was in the comment for 52ed2b4 earlier: "This is needed because some UI types have compound keys.". I called this out as a specific example of an earlier change description that had helpful (to me) information.

Earlier on you mentioned a SQL test failure workaround in your branch. From the branch description it seems like you're planning on checking this in with M5 or even M4. Will there be intermittent SQL test failures until that is checked in?

vanzin commented:
I don't think it's worth it to call out every design decision here. The feature set exists because it will be used by the code in the subsequent milestones; so if reviewing this you either need a bit of faith that there's a good reason for the feature to exist, or you need to go check out the rest of the code, which is why I posted a link to it in my first comment.

Again, feel free to ask specific questions. But don't be surprised if the answer is "it will be used in a subsequent milestone".

The reason for this is because I don't want to send a single PR that has > 20k line changes. Nobody is going to review that.

> Will there be intermittent SQL test failures until that is checked in?

No, as you can see, tests pass.

vanzin commented:
> Earlier on you mentioned a SQL test failure workaround in your branch

That's unrelated to this change, it was a bug introduced in a separate, completely unrelated change.

jsoltren commented:
Thanks @vanzin. I agree that it's tough to split up a change in such a way that it is easy to review. This approach - intermediate diffs in one PR, and a set of PRs for the whole feature - seems like a good compromise.

I would like to check out later sections before I +1 this, so, bear with me as I have a look and comment on a few other PRs. Again, no red flags yet, but this is a bunch of code to read and understand.

Commit:
SHS-NG M1: small fix in loop condition.

SparkQA commented:
**[Test build #77433 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/77433/testReport)** for PR 17902 at commit [`171f7fe`](https://github.com/apache/spark/commit/171f7fe6dc97132d57ca8e8c97b3f511c9eb6f55).

SparkQA commented:
**[Test build #77433 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/77433/testReport)** for PR 17902 at commit [`171f7fe`](https://github.com/apache/spark/commit/171f7fe6dc97132d57ca8e8c97b3f511c9eb6f55).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/77433/
Test PASSed.

SparkQA commented:
**[Test build #77561 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/77561/testReport)** for PR 17902 at commit [`c433065`](https://github.com/apache/spark/commit/c4330652ba8a6d053828ba358deafe0ad87422ee).

SparkQA commented:
**[Test build #77561 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/77561/testReport)** for PR 17902 at commit [`c433065`](https://github.com/apache/spark/commit/c4330652ba8a6d053828ba358deafe0ad87422ee).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/77561/
Test PASSed.

squito reviewed (commented):
I still have to look at the key schema a little bit more closely and the tests, but thought I'd checkpoint my feedback here.  Overall seems reasonable

vanzin commented on common/kvstore/src/main/java/org/apache/spark/kvstore/KVTypeInfo.java:
```diff
@@ -0,0 +1,154 @@
+
+    @Override
+    public Object get(Object instance) throws Exception {
+      return field.get(instance);
```
I thought I had included that change here but it ended up in another patch... will add it here instead.

vanzin commented on common/kvstore/src/main/java/org/apache/spark/kvstore/LevelDB.java:
```diff
@@ -0,0 +1,303 @@
+        }
+
+        PrefixCache cache = new PrefixCache(value);
+        byte[] naturalKey = ti.naturalIndex().toKey(ti.naturalIndex().getValue(value));
```
In LevelDBTypeInfo.java:

```
      Preconditions.checkNotNull(indexValue, "Null index value for %s in type %s.",
        name, type.getName());
```

Commit:
SHS-NG M1: feedback.

SparkQA commented:
**[Test build #77617 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/77617/testReport)** for PR 17902 at commit [`cc6662b`](https://github.com/apache/spark/commit/cc6662b54f25cab3274a3536c9bb722371278b88).

Commit:
SHS-NG M1: Fix LevelDBIterator.skip().

It was counting things incorrectly, and potentially returning the wrong
value; the change also defers deserialization to when next() is called,
so that hasNext() is cheaper.

SparkQA commented:
**[Test build #77617 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/77617/testReport)** for PR 17902 at commit [`cc6662b`](https://github.com/apache/spark/commit/cc6662b54f25cab3274a3536c9bb722371278b88).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/77617/
Test PASSed.

squito commented on common/kvstore/src/test/java/org/apache/spark/kvstore/LevelDBTypeInfoSuite.java:
```diff
@@ -0,0 +1,192 @@
+
+    public String id;
+
+  }
```
you also require that there is one element with a bare `@KVIndex`, with no `value`, to get a natural index, right?  I think its worth adding a case w/ valid `@KVIndex` but all have values.

squito reviewed (commented):
another round, couple more small comments.  Just need a bit more time looking at the tests, but still think it looks fine.

vanzin commented on common/kvstore/src/main/java/org/apache/spark/kvstore/LevelDBTypeInfo.java:
```diff
@@ -0,0 +1,502 @@
+ *
+ * <pre>
+ * Foo __main__ +key1   [data for instance 1]
+ * Foo __main__ +key2   [data for instance 2]
```
Yeah, it's a little confusing because I try to only call "key" the full key in LevelDB. Let me look at the comments and try to make them consistent, at least.

Commit:
SHS-NG M1: feedback 2.

SparkQA commented:
**[Test build #77651 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/77651/testReport)** for PR 17902 at commit [`2e809d3`](https://github.com/apache/spark/commit/2e809d3ad4e040d3d3168903e223965da2c65f2a).

SparkQA commented:
**[Test build #77651 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/77651/testReport)** for PR 17902 at commit [`2e809d3`](https://github.com/apache/spark/commit/2e809d3ad4e040d3d3168903e223965da2c65f2a).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/77651/
Test PASSed.

squito reviewed (commented):
I went through the tests and it looks pretty good.

After one more pass, I did have a high-level question: do you think its important to include jackson at this level?  You could just as easily leave the data serialization abstract.  The plus side is that that by leaving jackson out, this module becomes easier to include in other places (especially since jackson can be a mess with versioning).

or do you figure jackson is already widespread within in spark?

```
> find . -name "pom.xml" | xargs grep -m 1 "jackson"
./common/kvstore/pom.xml:      <groupId>com.fasterxml.jackson.core</groupId>
./common/network-common/pom.xml:      <groupId>com.fasterxml.jackson.core</groupId>
./common/network-yarn/pom.xml:              <pattern>com.fasterxml.jackson</pattern>
./core/pom.xml:      <artifactId>json4s-jackson_${scala.binary.version}</artifactId>
./external/kinesis-asl-assembly/pom.xml:      <groupId>com.fasterxml.jackson.core</groupId>
./hadoop-cloud/pom.xml:      the AWS module pulls in jackson; its transitive dependencies can create
./pom.xml:    <codehaus.jackson.version>1.9.13</codehaus.jackson.version>
./sql/core/pom.xml:      <groupId>com.fasterxml.jackson.core</groupId>
./sql/hive/pom.xml:      <groupId>org.codehaus.jackson</groupId>
```

vanzin commented:
> do you think its important to include jackson at this level?

I wouldn't say "important", but rather that it's a nice default. Jackson seems to be everywhere from core onwards, and even the other use I had for this (shuffle service, which I didn't change because it has some clever usage of keys that does not map to the code here well) uses Jackson.

I can easily remove Jackson and let anyone implement their own; but really the only implementation we'll have in Spark is Jackson-based (and some other, probably Java serialization, for the tests here), so it didn't feel like a big gain.

squito commented:
fair enough on jackson.

add the log for the seed please, other than that, lgtm

Commit:
SHS-NG M1: Log the test's random seed.

SparkQA commented:
**[Test build #77750 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/77750/testReport)** for PR 17902 at commit [`53d7d57`](https://github.com/apache/spark/commit/53d7d57196317ef406875f73d6e806f3b7956ee7).

SparkQA commented:
**[Test build #77750 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/77750/testReport)** for PR 17902 at commit [`53d7d57`](https://github.com/apache/spark/commit/53d7d57196317ef406875f73d6e806f3b7956ee7).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/77750/
Test PASSed.

squito commented:
merged to master

cloud-fan commented on common/kvstore/src/main/java/org/apache/spark/kvstore/KVStoreSerializer.java:
```diff
@@ -0,0 +1,86 @@
+  }
+
+  public final byte[] serialize(Object o) throws Exception {
+    if (o instanceof String) {
```
this only handles top-level string right? How about string field in a class?

vanzin replied:
Classes are handled by Jackson (the "else").

cloud-fan replied:
yea, but then we don't have the efficient UTF8 encoding, isn't it?

vanzin replied:
Why does that matter? Objects are written as JSON, it's not like the goal there is efficiency.

cloud-fan commented on common/kvstore/src/main/java/org/apache/spark/kvstore/KVStore.java:
```diff
@@ -0,0 +1,129 @@
+ * Data is also automatically compressed to save disk space.
+ * </p>
+ *
+ * <h3>Automatic Key Management</h3>
```
it will be better to have some examples, as I get confused here. What does `Automatic` mean? I thought users have to provide a natural index.

vanzin replied:
First, you do realize this PR has been merged a long time ago, right?

Automatic means you don't have to manually create keys. You're writing objects to the store, not calling something like `.put(key, object)`.

cloud-fan commented on common/kvstore/src/main/java/org/apache/spark/kvstore/LevelDBTypeInfo.java:
```diff
@@ -0,0 +1,516 @@
+ *   -NATURAL_INDEX
+ *   INDEX_NAME
+ *     +INDEX_VALUE
+ *       +NATURAL_KEY
```
is this the so-called pointer? For secondary index the value is natural key, which can be used to get the actual value/app-data.

cloud-fan commented on common/kvstore/src/main/java/org/apache/spark/kvstore/LevelDBTypeInfo.java:
```diff
@@ -0,0 +1,516 @@
+ * Foo __main__ +key1   [data for instance 1]
+ * Foo __main__ +key2   [data for instance 2]
+ * Foo __main__ -       [count of all Foo]
+ * Foo bar +yes +key1   [instance 1 key or data, depending on index type]
```
since `key1` is already in key, why we still need to put `key` in the value region?

vanzin replied:
Because otherwise, when reading from the index, it's not that easy to parse the object's key from the leveldb key so that you can retrieve the object itself.

You also have to store something, leveldb doesn't allow you to store null.

cloud-fan commented on common/kvstore/src/main/java/org/apache/spark/kvstore/LevelDBTypeInfo.java:
```diff
@@ -0,0 +1,516 @@
+ * ...
+ * Foo bar +yes -
+ * Foo bar .yes .child +no +key1   [instance 1 key or data, depending on index type]
+ * Foo bar .yes .child +no +key2   [instance 2 key or data, depending on index type]
```
One question, do we really use the user-given index name in leveldb key? Since we have the type info, we can map user-given index name to some unique ids and use it in leveldb key, to save space.

vanzin replied:
We could map the index name to a shorter identifier like it's done for types; but index names are generally short. And I've done some changes in code that uses this (not up for review yet) and shortening the index names doesn't end up saving a lot of space.

ConeyLiu commented on common/kvstore/src/main/java/org/apache/spark/kvstore/LevelDBTypeInfo.java:
```diff
@@ -0,0 +1,516 @@
+    private Index(KVIndex self, KVTypeInfo.Accessor accessor, Index parent) {
+      byte[] name = self.value().getBytes(UTF_8);
+      if (parent != null) {
+        byte[] child = new byte[name.length + 1];
```
Hi @vanzin,  what's meaning for this `child`?

vanzin replied:
Looks like dead code.
